Fix autolinking errors

ryzokuken opened the pull request:
This commit fixes some edge cases in the autolinking logic and adds tests for them. It fixes the issues with regexp, email validation as well as link annotations being injected multiple times on zoom.

Fixes: https://github.com/mozilla/pdf.js/issues/19462

ryzokuken committed:
Fix autolinking errors due to regex and email validation

Fix some edge cases in the autolinking logic with the regex as well as
validating email domains and add unit tests for them.

Fixes: https://github.com/mozilla/pdf.js/issues/19462

calixteman commented:
Ideally, it'd be nice to have a separate patch with an integration test about the annotation being injected multiple times, thank you.

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -257,6 +259,8 @@ class AnnotationLayerBuilder {
       this.div.hidden = false;
     }
+
+    this.#linksInjected = true;
```
The same line is should also be added to the early-return branch above https://github.com/mozilla/pdf.js/blob/e3cca6d513d4c24393cf5d8d4af25d69b399b46e/web/annotation_layer_builder.js#L246-L248 since otherwise there'd still be unnecessary re-parsing happening (even if nothing renders in that case).

ryzokuken replied:
Since the only other code path would be to go through with the entire rest of the function, what would be wrong with just moving the statement above this branch?

Snuffleupagus replied:
Right, it could just be placed directly after https://github.com/mozilla/pdf.js/pull/19470/files#diff-b166eae766e0c30a0a7dba9e750bd020d5ce9b432b84d6cb722aab3a8ca95aa1R242 instead.

ryzokuken committed:
Fix autolinking error due to redundant annotations on zooming

Fix an issue where redundant links were being added to the annotation
layer on zooming on the page with the links.

ryzokuken commented:
@calixteman broke out the zooming fix into its own patch and added integration test.

Snuffleupagus commented on test/integration/autolinker_spec.mjs:
```diff
@@ -121,4 +121,42 @@ describe("autolinker", function () {
+            annotations => annotations.map(a => a.href)
+          );
+          expect(url.length).withContext(`In ${browserName}`).toEqual(1);
+          await page.evaluate(() => window.PDFViewerApplication.zoomReset());
```
I believe that there's a couple of issues with this line:
 - It *implicitly* relies on the initial zoom level of the document being "page-fit", and if the default zoom-value for the integration-tests is ever changed this test would no longer be testing the intended thing, so it'd obviously be much better to invoke zoomIn or zoomOut here.
 - Just awaiting the zoom-call probably doesn't really guarantee that re-rendering has finished *or* that the annotationLayer has been re-enabled (since it's hidden during zooming). Hence you need to wait for the appropriate events, since otherwise the checks below might not be testing the correct thing.
 - Lastly, having a blank-line before this "second part" of the test-case wouldn't hurt.

nicolo-ribaudo replied:
@ryzokuken You can take a look at this test for how to wait for the zoom to happen: https://github.com/mozilla/pdf.js/blob/d6f63d0e4b35b31355bdf4f8338e284bf72c75b2/test/integration/viewer_spec.mjs#L231

Snuffleupagus replied:
Please also see PR #19475 which adds a new event that might be helpful here, since it'll allow detecting when the inferred-links have actually been added.

ryzokuken replied:
Great, thanks for your PR, will rebase on top once it's merged and listen on that event to avoid zooming in before the links are injected.

ryzokuken replied:
I believe all three of your points were addressed, so I'm marking this as resolved.

ryzokuken commented:
@Snuffleupagus @calixteman check out the PR now, the test waits for the link annotations and I think it should be looking much better/reliable now overall.

nicolo-ribaudo commented on test/integration/autolinker_spec.mjs:
```diff
@@ -121,4 +121,47 @@ describe("autolinker", function () {
+            annotations => annotations.map(a => a.href)
+          );
+          expect(url.length).withContext(`In ${browserName}`).toEqual(1);
+          await page.evaluate(() =>
```
```suggestion

          await page.evaluate(() =>
```
(https://github.com/mozilla/pdf.js/pull/19470#discussion_r1952272752)

ryzokuken replied:
er, sorry. I made a tiny mistake pushing my changes.

Snuffleupagus commented:
Taking another look at the existing test-cases, I'm not sure that any of them actually work *as intended* since they only wait for the `annotationLayer` to be available, which is however no guarantee that the inferred links have been processed.

ryzokuken commented:
@Snuffleupagus fair point, although I did verify locally that the tests work as intended but I agree that it's more reliable to wait for the event to be fired before checking. Added the waits before every `$$eval` now.

nicolo-ribaudo commented on test/integration/autolinker_spec.mjs:
```diff
@@ -121,4 +132,50 @@ describe("autolinker", function () {
+              scaleFactor: 2,
+            })
+          );
+          await waitForLinkAnnotations(page);
```
Does this actually work? If there are no link annotations added (because we already added them), the event is not going to be fired, right?

Should we maybe add a page after the one with the links, and wait for that page after to trigger `pagerendered`? (since it will only be rendered once the first page is done rendering all the layers)

Snuffleupagus replied:
The event *should* still fire (in the same the e.g. `annotationlayerrendered`, and many others, are fired when the annotationLayer has been zoomed).

Snuffleupagus commented:
/botio unittest

Snuffleupagus commented:
/botio integrationtest

ryzokuken commented:
All test failures are unrelated.

Snuffleupagus commented on web/autolinker.js:
```diff
@@ -96,21 +96,32 @@ class Autolinker {
+        raw = match[0];
+      } else {
+        // Check if the email address is valid through the URL constructor.
+        // In the future, this case be replaced by `URL.canParse`.
```
It seems that core-js has support for `URL.canParse` already, see https://github.com/zloirock/core-js?tab=readme-ov-file#url-and-urlsearchparams, so we might be able to use it directly?

calixteman commented on web/autolinker.js:
```diff
@@ -96,21 +96,32 @@ class Autolinker {
-      const raw =
+      let raw;
+      if (
         match[0].startsWith("www.") ||
```
nit: set a variable for `match[0]`.

calixteman reviewed (approved):
LGTM. Thank you.

Snuffleupagus commented on web/autolinker.js:
```diff
@@ -96,31 +96,38 @@ class Autolinker {
+        url.startsWith("https://")
+      ) {
+        raw = url;
+      } else {
+        if (!URL.canParse(`http://${emailDomain}`)) {
+          continue;
+        }
+        raw = url.startsWith("mailto:") ? url : `mailto:${url}`;
+      }
```
Can't this be simplified/improved a little bit, now that the `try-catch` block was removed?
```suggestion
      } else if (URL.canParse(`http://${emailDomain}`)) {
        raw = url.startsWith("mailto:") ? url : `mailto:${url}`;
      } else {
        continue;
      }     
```

ryzokuken replied:
Fixed, thanks!

Snuffleupagus commented:
/botio unittest

Snuffleupagus commented:
/botio integrationtest
